### PR TITLE
feat(builder): support project references for source code build mode

### PR DESCRIPTION
**Pull request by targeral**
### <samp>🤖 Generated by Copilot at 9438c1f</samp>

This pull request adds support for source code build mode for TypeScript projects with project references in the `@modern-js/builder` package. It also updates the integration tests and the changeset file for this feature.

### <samp>🤖 Generated by Copilot at 9438c1f</samp>

*  Add a changeset file to document the changes for the `@changesets/cli` tool ([link](https://github.com/web-infra-dev/modern.js/pull/4286/files?diff=unified&w=0#diff-6cf6d0610c6009a735393ffd136677c2331331db3c56e07404852c246bdecc1bR1-R6))
*  Import a type definition from `@modern-js/builder-webpack-provider` to access the webpack builder plugin configuration ([link](https://github.com/web-infra-dev/modern.js/pull/4286/files?diff=unified&w=0#diff-e894d618992ca3ef64b2f399f4275b6723da055ccb84fc13c8d8c9409a43f5e8R2))
*  Add a variable `enableSourceBuild` to check if the `experiments.sourceBuild` option is enabled in the builder configuration ([link](https://github.com/web-infra-dev/modern.js/pull/4286/files?diff=unified&w=0#diff-e894d618992ca3ef64b2f399f4275b6723da055ccb84fc13c8d8c9409a43f5e8R48-R52))
*  Add a conditional object to the `ForkTsCheckerWebpackPlugin` options to enable type checking for project references in source build mode ([link](https://github.com/web-infra-dev/modern.js/pull/4286/files?diff=unified&w=0#diff-e894d618992ca3ef64b2f399f4275b6723da055ccb84fc13c8d8c9409a43f5e8R61-R68))
*  Remove the `output.disableTsChecker` option from the `app` package configuration in the integration test for source code build, as it is no longer needed ([link](https://github.com/web-infra-dev/modern.js/pull/4286/files?diff=unified&w=0#diff-0f45805098e87aa19a1053e82cc534f1bc506d28b7f20706cd60cf08038f88dcL7-L9))
*  Fix a typo in the `card` entry point of the `components` package in the integration test for source code build, to point to the `.d.ts` file instead of the `.ts` file ([link](https://github.com/web-infra-dev/modern.js/pull/4286/files?diff=unified&w=0#diff-973cc6dbda9e51c378857f61583d8fea2ebd5a4a879dd0e2c3f3baa627446d3bL24-R24))

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/builder/builder/src/plugins/tsChecker.ts [hunk 1]
```diff
@@ -44,6 +45,11 @@ export const builderPluginTsChecker = (): DefaultBuilderPlugin => {
           return;
         }
 
+        const { experiments } = (
+          api as WebpackBuilderPluginAPI
+        ).getNormalizedConfig();
+        const enableSourceBuild = experiments?.sourceBuild ?? false;
```
Comment (chenjiahan) [inline packages/builder/builder/src/plugins/tsChecker.ts:51]: Normally we do not need to set a default value for a normalized config, this default value should be maintained in the default config.

Comment (targeral) [reply packages/builder/builder/src/plugins/tsChecker.ts:51]: Maybe there is no default when using `builder-rspack-provider`?

Comment (targeral) [reply packages/builder/builder/src/plugins/tsChecker.ts:51]: `const config = api.getNormalizedConfig();` 

With this code, I don't have access to the type definition of experiments.sourceBuild.

Comment (chenjiahan) [reply packages/builder/builder/src/plugins/tsChecker.ts:51]: I think it is bacuase `SharedNormalizedConfig` is missing `experiments` field, we can add it.

Comment (targeral) [reply packages/builder/builder/src/plugins/tsChecker.ts:51]: I'll try to add it, you can review it for me.

Comment (targeral) [reply packages/builder/builder/src/plugins/tsChecker.ts:51]: I found that the `sourceBuild` field cannot be added on the `SharedExperimentsConfig` type, as it is currently only available in the webpack builder.

Comment (targeral) [reply packages/builder/builder/src/plugins/tsChecker.ts:51]: So I think it's still necessary to determine if sourceBuild exists and provide a default value.

Comment (chenjiahan) [reply packages/builder/builder/src/plugins/tsChecker.ts:51]: Ok, the type can be improved after we support `sourceBuild` in Rspack provider.

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**